### ISSUE TITLE
Get Odds Data Automatically

### ISSUE THREAD

**Issue by kyleskom** [kyleskom/NBA-Machine-Learning-Sports-Betting#13]
We want to be able to get the odds data automatically without having to manually input data for games.

**Comment by mghattas100** [comment]
I wanted to know where you got the odds data from for all past seasons. odds of today's games for example are different in FanDuel vs. Draftkings. any recommendations which odds input you recommend for the user?

**Comment by kyleskom** [comment]
https://www.sportsbookreviewsonline.com/scoresoddsarchives/nba/nbaoddsarchives.htm

Any odds are fine! It'll calculate percentages and EV with respect to those.

**Comment by STRATZ-Ken** [comment]
> We want to be able to get the odds data automatically without having to manually input data for games.

I scraped this in a previous project.  Here is some python code.

You won't need all the code, and it will have to be adjusted for Money line but..


```
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
fanduel_odds_yes = {}
fanduel_odds_no = {}
driver = webdriver.Chrome(executable_path='chromedriver.exe')
driver.get("https://sportsbook.fanduel.com/navigation/nba")
regexSearch = re.findall("3[0-9]{7}\"", driver.page_source)
driver.close()
for r in regexSearch:
    id = r.replace("\"", "")
    URL = "https://sbapi.pa.sportsbook.fanduel.com/api/event-page?betexRegion=GBR&capiJurisdiction=intl&currencyCode=USD&exchangeLocale=en_US&includePrices=true&language=en&priceHistory=1&regionCode=NAMERICA&usePlayerPropsVirtualMarket=true&_ak=FhMFpcPWXMeyZxOx&eventId=%s" % (id)
    page = requests.get(URL)
    if page.status_code != 200:
        print('failed')
    parsed = json.loads(page.content)
    if "markets" in parsed["attachments"]:
        for market_id in parsed["attachments"]["markets"]:
            if parsed["attachments"]["markets"][market_id]["marketType"] == "2_OR_MORE_3_POINT_FGS_MADE_IN_THE_FIRST_3_MINUTES":
                fanduel_odds_yes[parsed["attachments"]["markets"][market_id]["marketName"].strip()] = (parsed["attachments"]["markets"][market_id]["runners"][0]["winRunnerOdds"]["americanDisplayOdds"]["americanOdds"])
                fanduel_odds_no[parsed["attachments"]["markets"][market_id]["marketName"].strip()] = (parsed["attachments"]["markets"][market_id]["runners"][1]["winRunnerOdds"]["americanDisplayOdds"]["americanOdds"])
print(fanduel_odds_yes)
for o, answer in enumerate(accuracy):
    if "2 or More 3 Point FGs Made in the First 3 Minutes - %s/%s" % (home_team_abbr[o], away_team_abbr[o]) in fanduel_odds_yes:
        print('Game : %s@%s /// FD Yes Odds : %s /// FD No Odds: %s /// Our Yes Odds: %s' % (
            away_team_abbr[o], 
            home_team_abbr[o], 
            fanduel_odds_yes["2 or More 3 Point FGs Made in the First 3 Minutes - %s/%s" % (home_team_abbr[o], away_team_abbr[o])], 
            fanduel_odds_no["2 or More 3 Point FGs Made in the First 3 Minutes - %s/%s" % (home_team_abbr[o], away_team_abbr[o])], 
            answer[0]
            ))
    elif "2 or More 3 Point FGs Made in the First 3 Minutes - %s/%s" % (away_team_abbr[o], home_team_abbr[o]) in fanduel_odds_yes:
        print('Game : %s@%s /// FD Yes Odds : %s /// FD No Odds: %s /// Our Yes Odds: %s' % (
            away_team_abbr[o], 
            home_team_abbr[o], 
            fanduel_odds_yes["2 or More 3 Point FGs Made in the First 3 Minutes - %s/%s" % (away_team_abbr[o], home_team_abbr[o])], 
            fanduel_odds_no["2 or More 3 Point FGs Made in the First 3 Minutes - %s/%s" % (away_team_abbr[o], home_team_abbr[o])], 
            answer[0]
            ))
    else:
        print('Skipping Game : %s@%s /// Prediction %s' % (away_team_abbr[o], home_team_abbr[o], answer[0]))
```

**Comment by Jojorabbit21** [comment]
Hello, @kyleskom .
It is my first time writing something into someone's repo.
First I want to thank you for creating this amazing pile of codes.

I've forked your repo and made scraping odds and games automatically with selenium.

I am not sure I've done it right, but I'd like to share my code here if it's okay.

forked repo: [https://github.com/Jojorabbit21/NBA-ML-PRED](url)

```python
# -*- coding: utf-8 -*-
import pandas as pd
import os.path

from src.Utils.Dictionaries import team_rotowire
from datetime import datetime
from pytz import timezone
from time import sleep
from selenium import webdriver
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys

def scrape_odds(date):
    
    url = 'https://www.rotowire.com/betting/nba/?date=' + date
    print(f"scraping_odds: {url}")
    
    user_agent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.47 Safari/537.36'
    options = webdriver.ChromeOptions()
    options.add_argument('disable-gpu')
    options.add_argument('headless')
    options.add_experimental_option('excludeSwitches', ['enable-logging'])

    driver = webdriver.Chrome(options=options, executable_path=r'./src/Scraper/chromedriver.exe')
    
    print(f"scraping_odds: OPENING BROWSER")
    driver.get(url)
    driver.implicitly_wait(2)
    sleep(2)

    visit = []
    home = []
    visit_odds = []
    home_odds = []
    ou = []
    
    # Team Name Selector
    # div.webix_ss_body > div.webix_ss_left > div > div > div:nth-child(1) > a > span.hide-until-md
    # Moneyline Odds Selector
    # div.webix_ss_body > div.webix_ss_center > div > div:nth-child(2)
    
    print(f"scraping_odds: FETCHING TEAMS...")
    teams = driver.find_elements(By.CSS_SELECTOR, 'div.webix_ss_body > div.webix_ss_left > div > div > div > a > span.hide-until-md')
    for i in range(len(teams)):
        if i%2==0:
            visit.append(team_rotowire[teams[i].text])
        else:
            home.append(team_rotowire[teams[i].text])
        
    print(f"scraping_odds: FETCHING MONEYLINE...")
    moneyline = driver.find_elements(By.CSS_SELECTOR, 'div.webix_ss_body > div.webix_ss_center > div > div:nth-child(2) > div')
    for i in range(len(moneyline)):
        if i%2==0:
            if moneyline[i].text != '-':
                visit_odds.append(moneyline[i].text)
            else:
                visit_odds.append(0)
        else:
            if moneyline[i].text != '-':
                home_odds.append(moneyline[i].text)
            else:
                home_odds.append(0)
    
    print(f"scraping_odds: FETCHING UNDER/OVER ODDS...")
    ouodds = driver.find_elements(By.CSS_SELECTOR, 'div.webix_ss_body > div.webix_ss_center > div > div:nth-child(6) > div')
    for i in range(len(ouodds)):
        if i%2 == 0:
            if ouodds[i].text != '-':
                val = str(ouodds[i].text)
                ou.append(val[2:])
            else:
                ou.append(0)
           
    driver.close()
    
    print(f"scraping_odds: BUILDING DATAFRAME...")
   
    df = pd.DataFrame()
    df["Visit"] = visit
    df['Home'] = home
    df['V_Odd'] = visit_odds
    df['H_Odd'] = home_odds
    df['OU'] = ou
    
    print(f"scraping_odds: SANITIZING DATAFRAME...")
    condition = (df.V_Odd == 0) | (df.H_Odd == 0) | (df.OU == 0)
    debris = df[condition].index
    df.drop(debris, inplace=True) 
    df.reset_index(drop=True, inplace=True)
    
    print(f"scraping_odds: DONE...")
    print(df)
    return df

```

This code scrapes specific day's game schedules and odds, then make it into dataframe and pass it to models to run.
As far, I think Rotowire is best option to scrape daily games and odds...

as a machine learning beginner, your codes are really helpful for me to learn.
Thank you again, @kyleskom !

**Comment by cainky** [comment]
What about using https://the-odds-api.com/ ?

**Comment by ayoub377** [comment]
can i use decimal odds or is it necessary to use american

**Comment by kyleskom** [comment]
@ayoub377 American

**Comment by cyberbee-cz** [comment]
Hi all, thank you first for creating this awesome project. I wanted to fill in odds data automatically so I implemented a little yahoo data scrapper which is parsing the json data structure and needs only requests and built-in json modules.

Here is scraper code sample: 
`
        import requests
    import json

    class YahooOddsProvider:
        data_url = 'https://sports.yahoo.com/nba/odds/'

        def __init__(self, data_url):
        self.data_url = data_url

        
        def get_odds(self):
            """Function returning odds from yahoo server's json content

            Returns:
                dictionary: [home_team_name + ':' + away_team_name: { home_team: money_line_odds, away_team: money_line_odds }, under_over_odds: val]
            """
            dict_res = {}
            response_data = requests.get(self.data_url)    
            startIndex = response_data.text.find('SportsbookStore') + len('SportsbookStore":')
            endIndex = response_data.text.find('ScoresStore') - 2
            sportsbook_data = json.loads(response_data.text[startIndex:endIndex])

            for game_name in sportsbook_data['state']['leagueOddsList']:
                game = sportsbook_data['state']['leagueOddsByGame'][game_name]

                # Get team names
                home_team_name = game['homeTeam']['location'] + ' ' + game['homeTeam']['nickname']
                away_team_name = game['awayTeam']['location'] + ' ' + game['awayTeam']['nickname']
                
                # Get money line bet values
                money_line_bet = next((x for x in game['bets'] if x['baseCategory'] == 'MONEY_LINE'), None)
                if money_line_bet is not None: 
                    money_line_home_value = next((x for x in money_line_bet['options'] if x['name'] == home_team_name))['americanOdds']
                    money_line_away_value = next((x for x in money_line_bet['options'] if x['name'] == away_team_name))['americanOdds']
                else: 
                    continue
                
                # Get totals bet value
                totals_bet = next((x for x in game['bets'] if x['baseCategory'] == 'TOTALS'), None)
                if totals_bet is not None: 
                    totals_value = float(totals_bet['options'][0]['optionDetails'][0]['value'])
                else: 
                    continue
                
                dict_res[home_team_name + ':' + away_team_name] =  { 
                    'under_over_odds': totals_value,
                    home_team_name: { 'money_line_odds': money_line_home_value }, 
                    away_team_name: { 'money_line_odds': money_line_away_value }
                }

            return dict_res
`

and related changes in main.py to be able to use new argument for data scraping. Feel free to use this code if you want to.

main.py 
`
    import argparse
    import pandas as pd
    import tensorflow as tf
    from src.Predict import NN_Runner, XGBoost_Runner
    from src.Utils.Dictionaries import team_index_current
    from src.Utils.tools import get_json_data, to_data_frame, get_todays_games_json, create_todays_games
    from src.DataProviders.YahooOddsProvider import YahooOddsProvider

    todays_games_url = 'https://data.nba.com/data/10s/v2015/json/mobile_teams/nba/2022/scores/00_todays_scores.json'
    data_url = 'https://stats.nba.com/stats/leaguedashteamstats?' \
            'Conference=&DateFrom=&DateTo=&Division=&GameScope=&' \
            'GameSegment=&LastNGames=0&LeagueID=00&Location=&' \
            'MeasureType=Base&Month=0&OpponentTeamID=0&Outcome=&' \
            'PORound=0&PaceAdjust=N&PerMode=PerGame&Period=0&' \
            'PlayerExperience=&PlayerPosition=&PlusMinus=N&Rank=N&' \
            'Season=2022-23&SeasonSegment=&SeasonType=Regular+Season&ShotClockRange=&' \
            'StarterBench=&TeamID=0&TwoWay=0&VsConference=&VsDivision='
    odds_provider = YahooOddsProvider('https://sports.yahoo.com/nba/odds/')

    def createTodaysGames(games, df, odds):
        match_data = []
        todays_games_uo = []
        home_team_odds = []
        away_team_odds = []

        for game in games:
            home_team = game[0]
            away_team = game[1]
            if odds is not None:
                game_odds = odds[home_team + ':' + away_team]
                todays_games_uo.append(game_odds['under_over_odds'])
                
                home_team_odds.append(game_odds[home_team]['money_line_odds'])
                away_team_odds.append(game_odds[away_team]['money_line_odds'])

            else:
                todays_games_uo.append(input(home_team + ' vs ' + away_team + ': '))

                home_team_odds.append(input(home_team + ' odds: '))
                away_team_odds.append(input(away_team + ' odds: '))

            home_team_series = df.iloc[team_index_current.get(home_team)]
            away_team_series = df.iloc[team_index_current.get(away_team)]
            stats = pd.concat([home_team_series, away_team_series])
            match_data.append(stats)

        games_data_frame = pd.concat(match_data, ignore_index=True, axis=1)
        games_data_frame = games_data_frame.T

        frame_ml = games_data_frame.drop(columns=['TEAM_ID', 'CFID', 'CFPARAMS', 'TEAM_NAME'])
        data = frame_ml.values
        data = data.astype(float)

        return data, todays_games_uo, frame_ml, home_team_odds, away_team_odds


    def main():
        odds = None
        data = get_todays_games_json(todays_games_url)
        games = create_todays_games(data)
        data = get_json_data(data_url)
        df = to_data_frame(data)
        if args.odds:
            print("--------------Automatic odds data scraping-------------")
            odds = odds_provider.get_odds()
        data, todays_games_uo, frame_ml, home_team_odds, away_team_odds = createTodaysGames(games, df, odds)
        if args.nn:
            print("------------Neural Network Model Predictions-----------")
            data = tf.keras.utils.normalize(data, axis=1)
            NN_Runner.nn_runner(data, todays_games_uo, frame_ml, games, home_team_odds, away_team_odds)
            print("-------------------------------------------------------")
        if args.xgb:
            print("---------------XGBoost Model Predictions---------------")
            XGBoost_Runner.xgb_runner(data, todays_games_uo, frame_ml, games, home_team_odds, away_team_odds)
            print("-------------------------------------------------------")
        if args.A:
            print("---------------XGBoost Model Predictions---------------")
            XGBoost_Runner.xgb_runner(data, todays_games_uo, frame_ml, games, home_team_odds, away_team_odds)
            print("-------------------------------------------------------")
            data = tf.keras.utils.normalize(data, axis=1)
            print("------------Neural Network Model Predictions-----------")
            NN_Runner.nn_runner(data, todays_games_uo, frame_ml, games, home_team_odds, away_team_odds)
            print("-------------------------------------------------------")


    if __name__ == "__main__":
        parser = argparse.ArgumentParser(description='Model to Run')
        parser.add_argument('-xgb', action='store_true', help='Run with XGBoost Model')
        parser.add_argument('-nn', action='store_true', help='Run with Neural Network Model')
        parser.add_argument('-A', action='store_true', help='Run all Models')
        parser.add_argument('-odds', action='store_true', help='Automatically parse in fill in current odds')
        args = parser.parse_args()
        main()
`
<img width="2187" alt="Screenshot 2022-12-19 at 23 39 46" src="https://user-images.githubusercontent.com/61155674/208541415-3d20ce1d-fb67-43ac-adb9-32255d1f12f7.png">

**Comment by kyleskom** [comment]
@cyberbee-cz Please open a pull request with this!

**Comment by cyberbee-cz** [comment]
@kyleskom here it is: #49

**Comment by kyleskom** [comment]
Done with #53
